port/osx/omrosbacktrace_impl.c

```diff
@@ -0,0 +1,411 @@
+	 * there are more frames but need to manually restore registers.
+	 */
+	unw_get_proc_name(&prev, buf, sizeof(buf), &ip);
+	if (0 == strncmp(buf, "_sigtramp", sizeof("_sigtramp"))) {
+		return handle_sigtramp(&prev, array, size, i);
+	}
+
+	return i;
+}
+
+/**
+ * NULL handler. We only care about preventing the signal from propagating up the call stack, no need to do
+ * anything in the handler.
+ */
+static uintptr_t
+handler(OMRPortLibrary *portLibrary, uint32_t gpType, void *gpInfo, void *userData)
+{
+	return OMRPORT_SIG_EXCEPTION_RETURN;
+}
+
+/**
+ * Wrapped call to libunwind backtrace functions
+ */
+uintptr_t
+protectedBacktrace(OMRPortLibrary *port, void *arg)
+{
+	frameData *addresses = (frameData *)arg;
+	return unw_backtrace(addresses->address_array, addresses->capacity);
+}
+
+/**
+ * Provides signal protection for the unw_backtrace call. If the stack walk causes a segfault then we check the output
+ * array to see if we got any frames as we may be able to provide a partial backtrace. We also record that the stack walk
+ * was incomplete so that information is available after the function returns.
+ */
+uintptr_t
+backtrace_sigprotect(OMRPortLibrary *portLibrary, J9PlatformThread *threadInfo, void **address_array, const int capacity)
+{
+	uintptr_t ret = 0;
+	frameData args;
+	args.address_array = address_array;
+	args.capacity = capacity;
+
+	memset(address_array, 0, sizeof(void *) * capacity);
+
+	if (omrthread_self()) {
+		if (0 != portLibrary->sig_protect(portLibrary, protectedBacktrace, &args, handler, NULL, OMRPORT_SIG_FLAG_SIGALLSYNC | OMRPORT_SIG_FLAG_MAY_RETURN, &ret)) {
+			/* check to see if there were any addresses populated */
+			for (ret = 0; ret < args.capacity && NULL != address_array[ret]; ret++);
+
+			threadInfo->error = FAULT_DURING_BACKTRACE;
+		}
+
+		return ret;
+	} else {
+		return unw_backtrace(address_array, capacity);
+	}
+}
+
+/**
+ * This function constructs a backtrace from a CPU context. Generally there are only one or two
+ * values in the context that are actually used to construct the stack but these vary by platform
+ * so arn't detailed here. If no heap is specified then this function will use malloc to allocate
+ * the memory necessary for the stack frame structures which must be freed by the caller.
+ *
+ * @param portLibrary a pointer to an initialized port library
+ * @param threadInfo the thread structure we want  to attach the backtrace to. Must not be NULL.
+ * @param heap a heap from which to allocate any necessary memory. If NULL malloc is used instead.
+ * @param signalInfo a platform signal context. If not null the context held in threadInfo is replaced
+ *  	  with the signal context before the backtrace is generated.
+ *
+ * @return the number of frames in the backtrace.
+ */
+uintptr_t
+omrintrospect_backtrace_thread_raw(OMRPortLibrary *portLibrary, J9PlatformThread *threadInfo, J9Heap *heap, void *signalInfo)
+{
+	void *addresses[50];
```

Comment:
This seems to limit the stack trace to 50 elements: what if the stack is deeper than that?

Reply:
I have been counting the number of elements during testing and I haven't seen any surpass 50. (Also I copied most of the two backtrace functions from the linux implementation, I assume any frames past 50 aren't as important)

Reply:
While further stack frames may generally less useful, it doesn't make them useless. We should aim to provide the entire stack (even if the Linux version doesn't (yet)).

Reply:
Should I increase it to the max (1024)? If there is a stack with that many frames, I think the heap allocated in javadump.cpp would run out of memory for the frames' symbols (see https://github.com/eclipse/omr/pull/6267#issuecomment-1005740252). More allocated `J9PlatformStackFrame`s -> more of the stack output as an address only? Unless we increase the size of the `backingStore` in javadump.cpp

Reply:
Tested 1024 by filling the addresses array with valid dummy addresses and with the current 8096 size backingStore I am only seeing 98 addresses in the javacore for each thread (no symbols). So even increasing the addresses array to size 98 we will not get any readable output in the javacore.

Reply:
I think the discussion about whether to capture more stack frames can wait. It's more important to at least have some.